lib/travis/build/script/shared/jdk.rb

```diff
@@ -4,7 +4,17 @@ class Script
       module Jdk
         def configure
           super
-          sh.cmd "jdk_switcher use #{config[:jdk]}", assert: true, echo: true, timing: false if uses_jdk?
+          if uses_jdk?
+            if use_install_jdk?(config[:jdk])
+              download_install_jdk
+              sh.if "-f install-jdk.sh" do
+                sh.cmd "source ./install-jdk.sh #{install_jdk_args config[:jdk]}", echo: true, assert: true
```

Comment:
Instead of `source`'ing, can this directly run `./install-jdk.sh {args}`?  I see that the project uses `source` in at least one place, but I'm assuming this is because something defined within is being used elsewhere, which I don't believe is the case for us.

Reply:
How are the updated environment variables, `JAVA_HOME` and `PATH` are set/modified by `install-jdk.sh`, propagated back to the calling shell then? Or am I missing something?

Reply:
I think the main function of `install_jdk.sh` is to download archives, and set `JAVA_HOME`. If `install_jdk.sh` can print this value out instead (say, with a flag?), we can use it; e.g.,

    export JAVA_HOME=$(./install-jdk.sh FLAGS)

Reply:
That would require every current user of `install-jdk.sh` to change the call like you suggested and take care of setting `JAVA_HOME` and adopt `PATH` accordingly. Those users will lose the _fire and forget_ feature, like in this _direct_ usage style: https://github.com/junit-team/junit5/blob/master/.travis.yml

At the moment `install-jdk.sh` already prints some information while running. Would that interfer with _can print this value out instead_? May bash-fu is limited here. PR is welcome. 😉

Reply:
See current print out here: https://travis-ci.org/sormuras/bach/jobs/362500362#L444

```
. ./install-jdk.sh -F 11 -L GPL
build.sh (2018-04-05)
  FEATURE = 11
  LICENSE = GPL
  ARCHIVE = openjdk-11-ea+7_linux-x64_bin.tar.gz
      URL = https://download.java.net/java/early_access/jdk11/7/GPL/openjdk-11-ea+7_linux-x64_bin.tar.gz [200]

--2018-04-05 07:49:43--  https://download.java.net/java/early_access/jdk11/7/GPL/openjdk-11-ea+7_linux-x64_bin.tar.gz
Resolving download.java.net (download.java.net)... 23.46.60.117
Connecting to download.java.net (download.java.net)|23.46.60.117|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 181183158 (173M) [application/x-gzip]
Saving to: â€˜openjdk-11-ea+7_linux-x64_bin.tar.gzâ€™


 0% [                                       ] 0           --.-K/s              
 5% [>                                      ] 9,251,383   41.6MB/s             
10% [===>                                   ] 19,297,463  44.7MB/s             
13% [====>                                  ] 25,174,016  38.8MB/s             
20% [======>                                ] 37,141,447  43.3MB/s             
26% [=========>                             ] 48,242,688  44.6MB/s             
32% [===========>                           ] 59,033,262  45.7MB/s             
38% [=============>                         ] 69,214,208  45.8MB/s             
44% [================>                      ] 79,727,934  46.2MB/s             
49% [==================>                    ] 90,185,728  46.5MB/s             
55% [====================>                  ] 100,671,488 46.5MB/s             
62% [=======================>               ] 113,254,400 47.4MB/s             
68% [=========================>             ] 123,740,160 47.4MB/s             
75% [============================>          ] 136,163,653 48.3MB/s             
81% [==============================>        ] 146,808,832 48.5MB/s             
86% [================================>      ] 157,294,592 48.6MB/s  eta 0s     
93% [===================================>   ] 169,176,509 49.0MB/s  eta 0s     
99% [=====================================> ] 180,363,264 49.8MB/s  eta 0s     
100%[======================================>] 181,183,158 49.9MB/s   in 3.5s   

2018-04-05 07:49:47 (49.3 MB/s) - â€˜openjdk-11-ea+7_linux-x64_bin.tar.gzâ€™ saved [181183158/181183158]

openjdk-11-ea+7_linux-x64_bin.tar.gz: gzip compressed data, from Unix, last modified: Fri Mar 30 03:14:13 2018
openjdk 11-ea 2018-09-18
OpenJDK Runtime Environment (build 11-ea+7)
OpenJDK 64-Bit Server VM (build 11-ea+7, mixed mode)
```

Would it suffice if I included a `JAVA_HOME = <...>` line?

Reply:
About to add an `echo ${JAVA_HOME}` as the last script command. With this, it possible to:

`JAVA_HOME=$(./install-jdk.sh -L GPL | tail -n 1)`

No more `source`-ing.

Reply:
Done via https://github.com/sormuras/bach/commit/d75685e693200cc0384972d8ff0497877385e4f7

Here is the result of `install-jdk-tests.sh`:
```java
install-jdk-tests.sh -- testing main variants in dry-run mode...

+ ./install-jdk.sh -D -L GPL -F 9
install-jdk.sh (2018-04-05)
  FEATURE = 9
  LICENSE = GPL
  ARCHIVE = openjdk-9.0.4_linux-x64_bin.tar.gz
      URL = https://download.java.net/java/GA/jdk9/9.0.4/binaries/openjdk-9.0.4_linux-x64_bin.tar.gz [200]

+ ./install-jdk.sh -D -L BCL -F 9
install-jdk.sh (2018-04-05)
  FEATURE = 9
  LICENSE = BCL
  ARCHIVE = jdk-9.0.4_linux-x64_bin.tar.gz
      URL = http://download.oracle.com/otn-pub/java/jdk/9.0.4+11/c2514751926b4512b076cc82f959763f/jdk-9.0.4_linux-x64_bin.tar.gz [302]

+ ./install-jdk.sh -D -L GPL -F 10
install-jdk.sh (2018-04-05)
  FEATURE = 10
  LICENSE = GPL
  ARCHIVE = openjdk-10_linux-x64_bin.tar.gz
      URL = https://download.java.net/java/GA/jdk10/10/binaries/openjdk-10_linux-x64_bin.tar.gz [200]

+ ./install-jdk.sh -D -L BCL -F 10
install-jdk.sh (2018-04-05)
  FEATURE = 10
  LICENSE = BCL
  ARCHIVE = jdk-10_linux-x64_bin.tar.gz
      URL = http://download.oracle.com/otn-pub/java/jdk/10+46/76eac37278c24557a3c4199677f19b62/jdk-10_linux-x64_bin.tar.gz [302]

+ ./install-jdk.sh -D -L GPL -F 11
install-jdk.sh (2018-04-05)
  FEATURE = 11
  LICENSE = GPL
  ARCHIVE = openjdk-11-ea+7_linux-x64_bin.tar.gz
      URL = https://download.java.net/java/early_access/jdk11/7/GPL/openjdk-11-ea+7_linux-x64_bin.tar.gz [200]

+ ./install-jdk.sh -D -L BCL -F 11
install-jdk.sh (2018-04-05)
  FEATURE = 11
  LICENSE = BCL
  ARCHIVE = jdk-11-ea+7_linux-x64_bin.tar.gz
      URL = https://download.java.net/java/early_access/jdk11/7/BCL/jdk-11-ea+7_linux-x64_bin.tar.gz [200]
```

Happy reviewing!